libkbfs/journal_server.go

```diff
@@ -359,11 +430,49 @@ func (j *JournalServer) JournalStatus(tlfID TlfID) (TLFJournalStatus, error) {
 	return tlfJournal.getJournalStatus()
 }
 
+// shutdownExistingJournals shuts down all write journals, sets the
+// current UID and verifying key to zero, and returns once all
+// shutdowns are complete. It is safe to call multiple times in a row,
+// and once this is called, EnableExistingJournals may be called
+// again.
+func (j *JournalServer) shutdownExistingJournals(ctx context.Context) {
+	j.log.CDebugf(ctx, "Shutting down existing journals")
+
+	j.lock.Lock()
+	defer j.lock.Unlock()
+	for _, tlfJournal := range j.tlfJournals {
+		tlfJournal.shutdown()
+	}
+
+	// Shutting down atomically is more important than respecting
+	// context cancellation.
+	waitCtx := context.Background()
+	for tlfID, tlfJournal := range j.tlfJournals {
+		err := tlfJournal.wait(waitCtx)
```

Comment:
I'm a bit worried about this, since in `tlfJournal` there could be work left in the work queue when it receives the shutdown signal, and the waitgroup might never go to zero as a result.  Maybe `tlfJournal` should flush the work channel and update the waitgroup accordingly on a hard shutdown?